pilot/pkg/networking/apigen/apigen.go

```diff
@@ -0,0 +1,165 @@
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/external"
+	"istio.io/istio/pkg/config/schema/collections"
+	"istio.io/istio/pkg/config/schema/resource"
+	"istio.io/pkg/log"
+)
+
+// APIGenerator supports generation of high-level API resources, similar with the MCP
+// protocol. This is a replacement for MCP, using XDS (and in future UDPA) as a transport.
+// Based on lessons from MCP, the protocol allows incremental updates by
+// default, using the same mechanism that EDS is using, i.e. sending only changed resources
+// in a push. Incremental deletes are sent as a resource with empty body.
+//
+// Example: networking.istio.io/v1alpha3/VirtualService
+//
+// TODO: we can also add a special marker in the header)
+type APIGenerator struct {
+}
+
+// TODO: take 'updates' into account, don't send pushes for resources that haven't changed
+// TODO: support WorkloadEntry - to generate endpoints (equivalent with EDS)
+// TODO: based on lessons from MCP, we want to send 'chunked' responses, like apiserver does.
+// A first attempt added a 'sync' record at the end. Based on feedback and common use, a
+// different approach can be used - for large responses, we can mark the last one as 'hasMore'
+// by adding a field to the envelope.
+
+// Generate implements the generate method for high level APIs, like Istio config types.
+// This provides similar functionality with MCP and :8080/debug/configz.
+//
+// Names are based on the current resource naming in istiod stores.
+func (g *APIGenerator) Generate(proxy *model.Proxy, push *model.PushContext, w *model.WatchedResource, updates model.XdsUpdates) model.Resources {
+	resp := []*golangany.Any{}
+
+	// Note: this is the style used by MCP and its config. Pilot is using 'Group/Version/Kind' as the
+	// key, which is similar.
+	//
+	// The actual type in the Any should be a real proto - which is based on the generated package name.
+	// For example: type is for Any is 'type.googlepis.com/istio.networking.v1alpha3.EnvoyFilter
+	// We use: networking.istio.io/v1alpha3/EnvoyFilter
+	gvk := strings.SplitN(w.TypeUrl, "/", 3)
+	if len(gvk) == 3 {
+		rgvk := resource.GroupVersionKind{
+			Group:   gvk[0],
+			Version: gvk[1],
+			Kind:    gvk[2],
+		}
+		if w.TypeUrl == collections.IstioMeshV1Alpha1MeshConfig.Resource().GroupVersionKind().String() {
+			meshAny, err := gogotypes.MarshalAny(push.Mesh)
+			if err == nil {
+				resp = append(resp, &golangany.Any{
+					TypeUrl: w.TypeUrl,
+					Value:   meshAny.Value,
+				})
+			}
+			return resp
+		}
+
+		// TODO: what is the proper way to handle errors ?
+		// Normally once istio is 'ready' List can't return errors on a valid config -
+		// even if k8s is disconnected, we still cache all previous results.
+		// This needs further consideration - I don't think XDS or MCP transports
+		// have a clear recommendation.
+		cfg, err := push.IstioConfigStore.List(rgvk, "")
```

Comment:
are we going to stream everything every N seconds (poll and get global state) ? No event interface to stream all configs only when there is change ?

Reply:
We'll certainly add that - but basic MCP in Galley sent all the configs at once, and I think it's better to split work into manageable chunks. 

I'm actually planning even more interesting improvements - like filtering out the configs based on caller ( part of the multi-tenant / isolation work that is discussed in env WG)

Reply:
Same concern. For ServiceEntry the response can be super huge.  This is far from production use.

Reply:
Istio 1.4 was used in production with identical behavior in Galley. 
I don't disagree that it's not ideal - and I mentioned the plans to fix it by using chunking, 
like apiserver does. 

However one PR can't do everything - this PR is only attempting to match 1.4 galley that 
we didn't finish merging into istiod in 1.5 and allow some experimentation with UDPA/XDS.
No plan to use it immediately in large scale production - we may as well remove like we removed duplicated chunks of galley.

Reply:
The behavior was not identical. I believe in MCP .List() is cheap and in the pilot CRD code its very expensive